testing/Android_SDK_E2E_Test suite.md

```diff
@@ -0,0 +1,71 @@
+
+### Webdriverio (nodejs)
+
+Like the WD.js library like the WD.js client, tests using this client can be written in a short chained syntax and can also be used in conjunction with features of tools like mochas transfer promisness to make assertions part of this chained syntax.  Unlike the WD.js client however, webdriverio can handle the redirect to the browser from the application, however there is only two possible locator strategies - xpath and accessibilityId. 
+
+### Java-Client (Appium Java SDK)
+
+Unlike the javascript libraries the java-client cannot be written in a promise chain syntax, however, there is a large number of locator strategies and it can handle the transition between native application and mobile browser.  Another major feature of the java client is the support for page object classes - using the following annotations, @FindBy(someStrategy), @AndroidFindBy(someStrategy) and @iOSFindBy(someStrategy), allows the creation of a single page object when creating tests for applications on multiple platforms that have the same content and flow but identifiers unique to their platform rather than a separate page object for each [3]. 
+
+## Framework Recommendation
+
+Appium is the framework of choice from those explored as it is the only framework explored that provides clients that can handle the redirect to the browser that is an integral test case we must cover.  Of the two clients, the Java-Client is likely to be the superior choice as it has a greater number of locator strategies and the annotation selector targeting provided can reduce the size of the codebase.
+
+## Test suite Proposal
+
+### Implementation Language
+
+With the selection of the Appium framework and its Java-Client java, or a JVM based language such as Kotlin, Groovy, Scala, etc. should be used, provided they can support the java based annotations that the appium Java-Client provides for Page Object support.  The use of a JVM language other than Java could increase the readability of the test suite.
+
+### Page Objects
+
+The test suite should make use of the page object model. This model separates the logic the presentation logic of the associated page objects and the test logic which leads to tests that are less likely to be brittle, and would therefore increase the maintainability of the test suite [4].  As mentioned in the framework recommendation section, the ability of the Appium library to use annotations for targeting selectors that are unique to each platform can help to reduce the codebase size - only one page object is required per platform with the annotations providing the ability to target Android or iOS specific selectors and the tests files themselves can also be generic meaning no separate tests for Android or iOS applications that share the exact same pattern of logic flow through the application.
+
+### Self-Contained Repository
```

Comment:
So if the tests are in a separate repo, when they will be triggered to run? How can SDK developers get quick feedbacks if something  is broken?

Reply:
@wei-lee What Appium requires is a path to the produced app file so it can upload it to the emulator and test it. This path is specified within the options passed to the appium client when running a test.

On a developer machine we could use a file the the location of build directories of multiple SDKs which will provide this information to the appium configuration options section. The steps to run the e2e tests then would be:

A. Trigger build from SDK repo
B. Trigger UI test command from e2e testing repo

Of course a shell script or system alias could be used to trigger these commands sequentially without the developer needing to manually trigger them both.